### PR TITLE
usb: Refactor to allow multiple classes

### PR DESCRIPTION
**Pull request by arturo182**
Good luck reviewing this :P

### CODE REVIEW

**Comment by kbeckmann** [comment]
37 files changed lol

File: app/drivers/pcd.c [hunk 1]
```diff
@@ -30,34 +34,145 @@ void HAL_PCD_MspDeInit(PCD_HandleTypeDef* p_pcd)
...
+	self.p_usbd->request.bRequest  = *(uint8_t *)(p_setup +  1);
+	self.p_usbd->request.wValue    =     SWAPBYTE(p_setup +  2);
+	self.p_usbd->request.wIndex    =     SWAPBYTE(p_setup +  4);
+	self.p_usbd->request.wLength   =     SWAPBYTE(p_setup +  6);
+
+	self.p_usbd->ep0_state = USBD_EP0_SETUP;
+	self.p_usbd->ep0_data_len = self.p_usbd->request.wLength;
+
+	switch (self.p_usbd->request.bmRequest & 0x1F) {
+		case USB_REQ_RECIPIENT_DEVICE:
```
Comment (kbeckmann) [inline app/drivers/pcd.c]: ah, so we're doing 
```
switch () {
	case foo:
		code();
		break;
}
```

File: app/drivers/usb.c [hunk 2]
```diff
@@ -18,11 +15,28 @@ static struct {
...
+#define USBD_PRODUCT_STRING_FS			"XilDebug CMSIS-DAP"
+#define USBD_SERIALNUMBER_STRING_FS		"123456789ABC"
+#define CDC_CTRL_INTERFACE_NO			0
+#define CDC_DATA_INTERFACE_NO			1
+#define HID_INTERFACE_NO				2
+
+#define USBD_USER_STRING_CDC			"CMSIS-DAP CDC"
+#define USBD_USER_STRING_CDC_IDX		4
+
+#define USBD_USER_STRING_DCI			"CMSIS-DAP DCI"
```
Comment (kbeckmann) [inline app/drivers/usb.c]: What is DCI?

File: app/drivers/pcd.c (hunk 1 above)
Comment (arturo182) [reply app/drivers/pcd.c]: Nope, my bad :P

File: app/drivers/usb.c (hunk 2 above)
Comment (arturo182) [reply app/drivers/usb.c]: Data Class Interface https://doc.micrium.com/display/DOC/CDC+Class+Overview

**Pull request merged**